File: sdk/python/muzen/tools.py
from __future__ import annotations

import asyncio
import inspect
import json
import re
import threading
import types
from dataclasses import MISSING, dataclass, fields, is_dataclass
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import (
    Any,
    Callable,
    Dict,
    List,
    Optional,
    Set,
    Type,
    Union,
    get_args,
    get_origin,
    get_type_hints,
)

from .agent import MuzenError


_TOOL_NAME = re.compile(r"^[a-zA-Z0-9_-]{1,64}$")


@dataclass(frozen=True)
class Tool:
    """A Python function exposed to an Agent as a local MCP tool."""

    function: Callable[..., Any]
    name: str
    description: str
    input_schema: Dict[str, Any]

    def invoke(self, arguments: Dict[str, Any]) -> Any:
        value = self.function(**arguments)
        if inspect.isawaitable(value):
            return asyncio.run(value)
        return value


def tool(function: Callable[..., Any]) -> Tool:
    """Wrap a typed sync or async function as a Muzen tool."""
    if isinstance(function, Tool):
        return function
    if not inspect.isfunction(function):
        raise _invalid("tool", "must be a function")
    name = function.__name__
    if _TOOL_NAME.fullmatch(name) is None:
        raise _invalid(
            "tool.name", "must match [a-zA-Z0-9_-]{1,64}"
        )
    return Tool(
        function=function,
        name=name,
        description=_first_docstring_paragraph(function),
        input_schema=_function_schema(function),
    )


class LoopbackToolServer:
    """Small Streamable HTTP MCP server for one Agent's Python tools."""

    def __init__(self, tools: tuple[Tool, ...]) -> None:
        self._tools = {item.name: item for item in tools}
        self._server: Optional[ThreadingHTTPServer] = None
        self._thread: Optional[threading.Thread] = None

    @property
    def url(self) -> str:
        if self._server is None:
            raise RuntimeError("tool server has not been started")
        return "http://127.0.0.1:%d/mcp" % self._server.server_port

    def start(self) -> None:
        if self._server is not None:
            return
        tools = self._tools

        class Handler(BaseHTTPRequestHandler):
            def do_POST(self) -> None:
                if self.path != "/mcp":
                    self.send_error(404)
                    return
                try:
                    length = int(self.headers.get("Content-Length", "0"))
                    request = json.loads(self.rfile.read(length).decode("utf-8"))
                except (ValueError, UnicodeDecodeError, json.JSONDecodeError):
                    self._json_error(None, -32700, "invalid JSON")
                    return
                request_id = request.get("id")
                method = request.get("method")
                if method == "notifications/initialized":
                    self.send_response(202)
                    self.send_header("Content-Length", "0")
                    self.end_headers()
                    return
                if method == "initialize":
                    self._json_result(
                        request_id,
                        {
                            "protocolVersion": "2025-03-26",
                            "capabilities": {"tools": {}},
                            "serverInfo": {"name": "muzen-python-tools", "version": "1"},
                        },
                        session_id="muzen-python-tools",
                    )
                    return
                if method == "tools/list":
                    listed = [
                        {
                            "name": item.name,
                            "description": item.description,
                            "inputSchema": item.input_schema,
                        }
                        for item in tools.values()
                    ]
                    self._json_result(request_id, {"tools": listed})
                    return
                if method == "tools/call":
                    params = request.get("params") or {}
                    selected = tools.get(params.get("name"))
                    if selected is None:
                        self._json_error(request_id, -32602, "unknown tool")
                        return
                    arguments = params.get("arguments", {})
                    if not isinstance(arguments, dict):
                        self._json_error(request_id, -32602, "arguments must be an object")
                        return
                    try:
                        value = selected.invoke(arguments)
                        encoded = json.dumps(value)
                        text = value if isinstance(value, str) else encoded
                        result: Dict[str, Any] = {
                            "content": [{"type": "text", "text": text}],
                            "isError": False,
                        }
                        if not isinstance(value, str):
                            result["structuredContent"] = value
                    except Exception as exc:
                        result = {
                            "content": [{"type": "text", "text": str(exc)}],
                            "isError": True,
                        }
                    self._json_result(request_id, result)
                    return
                self._json_error(request_id, -32601, "method not found")

            def _json_result(
                self, request_id: Any, result: Dict[str, Any], session_id: Optional[str] = None
            ) -> None:
                self._write_json(
                    {"jsonrpc": "2.0", "id": request_id, "result": result},
                    session_id,
                )

            def _json_error(self, request_id: Any, code: int, message: str) -> None:
                self._write_json(
                    {
                        "jsonrpc": "2.0",
                        "id": request_id,
                        "error": {"code": code, "message": message},
                    }
                )

            def _write_json(
                self, value: Dict[str, Any], session_id: Optional[str] = None
            ) -> None:
                body = json.dumps(value).encode("utf-8")
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.send_header("Content-Length", str(len(body)))
                if session_id is not None:
                    self.send_header("Mcp-Session-Id", session_id)
                self.end_headers()
                self.wfile.write(body)

            def log_message(self, format: str, *args: Any) -> None:
                pass

        self._server = ThreadingHTTPServer(("127.0.0.1", 0), Handler)
        self._thread = threading.Thread(
            target=self._server.serve_forever,
            name="muzen-python-tools",
            daemon=True,
        )
        self._thread.start()

    def close(self) -> None:
        if self._server is None:
            return
        self._server.shutdown()
        self._server.server_close()
        if self._thread is not None:
            self._thread.join(timeout=2)
        self._server = None
        self._thread = None


def annotation_schema(annotation: Any, active: Set[Type[Any]], path: str) -> Dict[str, Any]:
    primitive = {str: "string", int: "integer", float: "number", bool: "boolean"}
    if annotation in primitive:
        return {"type": primitive[annotation]}
    if isinstance(annotation, type) and (_is_typed_dict(annotation) or is_dataclass(annotation)):
        return _object_schema(annotation, active, path)
    origin = get_origin(annotation)
    args = get_args(annotation)
    if origin in (list, List):
        if len(args) != 1:
            raise _invalid(path, "list annotations must have one item type")
        return {"type": "array", "items": annotation_schema(args[0], active, path)}
    union_origins = [Union]
    union_type = getattr(types, "UnionType", None)
    if union_type is not None:
        union_origins.append(union_type)
    if origin in tuple(union_origins) and len(args) == 2 and type(None) in args:
        item = args[0] if args[1] is type(None) else args[1]
        return {"anyOf": [annotation_schema(item, active, path), {"type": "null"}]}
    raise _invalid(path, "contains unsupported annotation %r" % (annotation,))


def _function_schema(function: Callable[..., Any]) -> Dict[str, Any]:
    try:
        hints = get_type_hints(function)
    except (NameError, TypeError) as exc:
        raise _invalid("tool.%s" % function.__name__, "contains an unresolved annotation: %s" % exc)
    properties: Dict[str, Any] = {}
    required = []
    for parameter in inspect.signature(function).parameters.values():
        path = "tool.%s.%s" % (function.__name__, parameter.name)
        if parameter.kind in (parameter.VAR_POSITIONAL, parameter.VAR_KEYWORD):
            raise _invalid(path, "variadic parameters are unsupported")
        if parameter.kind == parameter.POSITIONAL_ONLY:
            raise _invalid(path, "positional-only parameters are unsupported")
        if parameter.name not in hints:
            raise _invalid(path, "must have a supported type annotation")
        properties[parameter.name] = annotation_schema(hints[parameter.name], set(), path)
        if parameter.default is inspect.Parameter.empty:
            required.append(parameter.name)
    schema: Dict[str, Any] = {
        "type": "object",
        "properties": properties,
        "additionalProperties": False,
    }
    if required:
        schema["required"] = required
    return schema


def _object_schema(value: Type[Any], active: Set[Type[Any]], path: str) -> Dict[str, Any]:
    if value in active:
        raise _invalid(path, "recursive annotations are unsupported")
    active.add(value)
    try:
        hints = get_type_hints(value)
        properties = {
            name: annotation_schema(annotation, active, path)
            for name, annotation in hints.items()
        }
        if _is_typed_dict(value):
            required_keys = getattr(value, "__required_keys__", None)
            required = list(hints) if required_keys is None else [
                name for name in hints if name in required_keys
            ]
        else:
            field_by_name = {item.name: item for item in fields(value)}
            required = [
                name for name in hints
                if field_by_name[name].default is MISSING
                and field_by_name[name].default_factory is MISSING
            ]
    except (NameError, TypeError) as exc:
        raise _invalid(path, "contains an unresolved annotation: %s" % exc)
    finally:
        active.remove(value)
    schema: Dict[str, Any] = {
        "type": "object",
        "properties": properties,
        "additionalProperties": False,
    }
    if required:
        schema["required"] = required
    return schema


def _first_docstring_paragraph(function: Callable[..., Any]) -> str:
    doc = inspect.getdoc(function) or ""
    paragraph = doc.split("\n\n", 1)[0]
    return " ".join(line.strip() for line in paragraph.splitlines())


def _is_typed_dict(value: Any) -> bool:
    return (
        isinstance(value, type)
        and issubclass(value, dict)
        and hasattr(value, "__annotations__")
        and hasattr(value, "__total__")
    )


def _invalid(path: str, message: str) -> MuzenError:
    return MuzenError("invalid_input", "%s %s" % (path, message), False, {"path": path})

File: sdk/typescript/packages/muzen-sdk/src/tools.ts
import { createServer, type IncomingMessage, type Server, type ServerResponse } from "node:http";

import { MuzenError, type JsonObject } from "./agent.js";

const TOOL_NAME = /^[a-zA-Z0-9_-]{1,64}$/;
const SERVER_NAME = "muzen-python-tools";

export interface ToolInputSchema extends JsonObject {
  type: "object";
  properties: Record<string, unknown>;
  required?: string[];
  additionalProperties: false;
}

export interface Tool<TArguments extends object = JsonObject> {
  readonly name: string;
  readonly description: string;
  readonly input: ToolInputSchema;
  readonly execute: (arguments_: TArguments) => unknown | Promise<unknown>;
}

export interface FunctionToolOptions {
  readonly description: string;
  readonly input: ToolInputSchema;
}

export function tool<TArguments extends object>(definition: Tool<TArguments>): Tool<TArguments>;
export function tool<TArguments extends object>(
  execute: (arguments_: TArguments) => unknown | Promise<unknown>,
  options: FunctionToolOptions,
): Tool<TArguments>;
export function tool<TArguments extends object>(
  definitionOrExecute: Tool<TArguments> | ((arguments_: TArguments) => unknown | Promise<unknown>),
  options?: FunctionToolOptions,
): Tool<TArguments> {
  if (typeof definitionOrExecute !== "function" && !isObject(definitionOrExecute)) {
    throw invalid("tool", "must be an object or function");
  }
  const definition = typeof definitionOrExecute === "function"
    ? {
        name: definitionOrExecute.name,
        description: options?.description,
        input: options?.input,
        execute: definitionOrExecute,
      }
    : definitionOrExecute;

  if (typeof definition.name !== "string" || !TOOL_NAME.test(definition.name)) {
    throw invalid("tool.name", "must match [a-zA-Z0-9_-]{1,64}");
  }
  if (typeof definition.description !== "string") {
    throw invalid(`tool.${definition.name}.description`, "must be a string");
  }
  validateInputSchema(definition.input, `tool.${definition.name}.input`);
  if (typeof definition.execute !== "function") {
    throw invalid(`tool.${definition.name}.execute`, "must be a function");
  }

  return Object.freeze({
    name: definition.name,
    description: definition.description,
    input: definition.input,
    execute: definition.execute,
  });
}

export class LoopbackToolServer {
  private readonly tools: Map<string, Tool<any>>;
  private server: Server | undefined;
  private startPromise: Promise<void> | undefined;

  constructor(tools: readonly Tool<any>[]) {
    this.tools = new Map(tools.map((item) => [item.name, item]));
  }

  get url(): string {
    const address = this.server?.address();
    if (address === null || address === undefined || typeof address === "string") {
      throw new Error("tool server has not been started");
    }
    return `http://127.0.0.1:${address.port}/mcp`;
  }

  async start(): Promise<void> {
    if (this.server?.listening === true) return;
    this.startPromise ??= this.listen();
    try {
      await this.startPromise;
    } catch (error) {
      this.startPromise = undefined;
      this.server = undefined;
      throw error;
    }
  }

  async close(): Promise<void> {
    if (this.startPromise !== undefined) {
      try {
        await this.startPromise;
      } catch {
        // A failed start has no listening server to close.
      }
    }
    const server = this.server;
    this.server = undefined;
    this.startPromise = undefined;
    if (server === undefined || !server.listening) return;
    await new Promise<void>((resolve, reject) => {
      server.close((error) => error === undefined ? resolve() : reject(error));
      server.closeIdleConnections();
    });
  }

  private async listen(): Promise<void> {
    const server = createServer((request, response) => void this.handle(request, response));
    this.server = server;
    await new Promise<void>((resolve, reject) => {
      const onError = (error: Error): void => {
        server.off("listening", onListening);
        reject(error);
      };
      const onListening = (): void => {
        server.off("error", onError);
        resolve();
      };
      server.once("error", onError);
      server.once("listening", onListening);
      server.listen(0, "127.0.0.1");
    });
  }

  private async handle(request: IncomingMessage, response: ServerResponse): Promise<void> {
    if (request.method !== "POST" || request.url !== "/mcp") {
      response.writeHead(404);
      response.end();
      return;
    }

    let payload: unknown;
    try {
      payload = JSON.parse(await readBody(request)) as unknown;
    } catch {
      writeError(response, null, -32700, "invalid JSON");
      return;
    }
    if (!isObject(payload)) {
      writeError(response, null, -32700, "invalid JSON");
      return;
    }

    const requestId = payload.id ?? null;
    const method = payload.method;
    if (method === "notifications/initialized") {
      response.writeHead(202, { "Content-Length": "0" });
      response.end();
      return;
    }
    if (method === "initialize") {
      writeResult(response, requestId, {
        protocolVersion: "2025-03-26",
        capabilities: { tools: {} },
        serverInfo: { name: SERVER_NAME, version: "1" },
      }, SERVER_NAME);
      return;
    }
    if (method === "tools/list") {
      writeResult(response, requestId, {
        tools: [...this.tools.values()].map((item) => ({
          name: item.name,
          description: item.description,
          inputSchema: item.input,
        })),
      });
      return;
    }
    if (method === "tools/call") {
      const params = isObject(payload.params) ? payload.params : {};
      const selected = typeof params.name === "string" ? this.tools.get(params.name) : undefined;
      if (selected === undefined) {
        writeError(response, requestId, -32602, "unknown tool");
        return;
      }
      const arguments_ = params.arguments === undefined ? {} : params.arguments;
      if (!isObject(arguments_)) {
        writeError(response, requestId, -32602, "arguments must be an object");
        return;
      }
      try {
        const value = await selected.execute(arguments_);
        const result: JsonObject = {
          content: [{ type: "text", text: typeof value === "string" ? value : pythonJsonStringify(value) }],
          isError: false,
        };
        if (typeof value !== "string") result.structuredContent = value;
        writeResult(response, requestId, result);
      } catch (error) {
        writeResult(response, requestId, {
          content: [{ type: "text", text: errorText(error) }],
          isError: true,
        });
      }
      return;
    }
    writeError(response, requestId, -32601, "method not found");
  }
}

function validateInputSchema(value: unknown, path: string): asserts value is ToolInputSchema {
  if (!isObject(value)) {
    throw invalid(path, "must be an object JSON Schema with properties and additionalProperties false");
  }
  const properties = value.properties;
  if (value.type !== "object" || !isObject(properties) || value.additionalProperties !== false) {
    throw invalid(path, "must be an object JSON Schema with properties and additionalProperties false");
  }
  if (
    value.required !== undefined &&
    (!Array.isArray(value.required) || value.required.some((item) => typeof item !== "string"))
  ) {
    throw invalid(`${path}.required`, "must be an array of property names");
  }
  if (
    Array.isArray(value.required) &&
    value.required.some((item) => typeof item === "string" && !(item in properties))
  ) {
    throw invalid(`${path}.required`, "must contain only declared property names");
  }
}

async function readBody(request: IncomingMessage): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const chunk of request) chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  return Buffer.concat(chunks).toString("utf8");
}

function writeResult(response: ServerResponse, requestId: unknown, result: unknown, sessionId?: string): void {
  writeJson(response, { jsonrpc: "2.0", id: requestId, result }, sessionId);
}

function writeError(response: ServerResponse, requestId: unknown, code: number, message: string): void {
  writeJson(response, { jsonrpc: "2.0", id: requestId, error: { code, message } });
}

function writeJson(response: ServerResponse, value: unknown, sessionId?: string): void {
  const body = JSON.stringify(value);
  response.writeHead(200, {
    "Content-Type": "application/json",
    "Content-Length": String(Buffer.byteLength(body)),
    ...(sessionId === undefined ? {} : { "Mcp-Session-Id": sessionId }),
  });
  response.end(body);
}

function pythonJsonStringify(value: unknown): string {
  const compact = JSON.stringify(value);
  if (compact === undefined) throw new TypeError("Object of type undefined is not JSON serializable");
  let result = "";
  let quoted = false;
  let escaped = false;
  for (const character of compact) {
    result += character;
    if (quoted) {
      if (escaped) escaped = false;
      else if (character === "\\") escaped = true;
      else if (character === '"') quoted = false;
    } else if (character === '"') {
      quoted = true;
    } else if (character === "," || character === ":") {
      result += " ";
    }
  }
  return result;
}

function errorText(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function isObject(value: unknown): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function invalid(path: string, message: string): MuzenError {
  return new MuzenError("invalid_input", `${path} ${message}`, false, { path });
}
